src/ekore/harmonics/cache.py

```diff
@@ -117,6 +130,30 @@ def get(key: int, cache: npt.ArrayLike,
     elif key == Sm211:
         s = Sm211(n, get(S1, cache, n), get(S2, cache, n), 
             get(Sm1, cache, n), is_singlet)
+    elif key == S1h:
+        s = w1.S1(n/2)
+    elif key == S2h:
+        s = w2.S2(n/2)
+    elif key == S3h:
+        s = w3.S3(n/2)
+    elif key == S1mh:
+        s = w1.S1((n-1)/2)
+    elif key == S2mh:
+        s = w2.S2((n-1)/2)
+    elif key == S3mh:
+        s = w3.s3((n-1)/2)
+    elif key == S1ph:
+        s = w1.S1((n+1)/2)
+    elif key == S2ph:
+        s = w2.S2((n+1)/2)
+    elif key == S3ph:
+        s = w3.S3((n+1)/2)
+    elif key == g3:
+        s = mellin_g3(n, get(S1, cache, n))
+    elif key == S1p2:
```

Comment:
$S1(N+2)$ can be obtained from $S1(N)$ with the recursive relations, i.e.
```
S1p2 = polygamma.recursive_harmonic_sum(cache[S1], n, 2, 1)
```
Moreover, once that you have g3p2 you can obtain for free `Sm21` using that
```
g3p2 = -1 / eta * (zeta2 * sm1_np1 - 5.0 / 8 * zeta3 + zeta2 * log2 - sm21_np1)
``` 
with
```
s1_np1 = polygamma.recursive_harmonic_sum(s1, n, 1, 1)
sm1_np1 = sm1 - eta / (n + 1)
sm21_np1 = sm21 - eta / (n + 1) ** 2 * s1_np1
```
and `eta` is the symmetry factor (i.e. 1 for singlet and -1 for non-singlet)

Reply:
I have made all the changes except the Sm21. As far as I can tell, it uses g3p1 (https://github.com/NNPDF/eko/blob/master/src/ekore/harmonics/w3.py#L138), however the comment (https://github.com/NNPDF/eko/blob/master/src/ekore/harmonics/w3.py#L135) is unclear to me. If it is instead using g3p2 (due to N-1 implementation), then perhaps it is inconsistent with (https://github.com/NNPDF/eko/blob/master/src/ekore/anomalous_dimensions/unpolarized/space_like/as1aem1.py#L263).
As far as I can tell, there are no longer any repeated harmonic sum calculations (for the ones that are implemented in the cache)

Reply:
Yes I know that it uses $g_3(N+1)$ and not $g_3(N+2)$ but the two are related through recursive relations (similar for the ones relating $S_1(N)$ with $S_1(N+1)$ but slightly more complicated). The equation implemented in https://github.com/NNPDF/eko/blob/master/src/ekore/harmonics/w3.py#L138 should be equivalent to the one in this comment (if my computation is correct).
The point here is that it is better to reuse the value of $g_3(N+2)$, that we already have (actually this is true only if QED is switched on), instead of recomputing $g_3(N+1)$.